### PR TITLE
 Added config option that allows to preserve zero margins

### LINKED ISSUE #2935: Paste From Word (PFW) filter removes margin-top, margin-bottom
## Type of report

Feature request

## Provide detailed reproduction steps (if any)

1. Open attached Word file
2. Disable ACF in your editor.
2. Select All and paste into editor 
[Margin Top.zip](https://github.com/ckeditor/ckeditor-dev/files/2941726/Margin.Top.zip)

### Expected result

Spacing between lines are small just like in MS Word

### Actual result

Spacing between lines are huge.

## Other details

* Browser: Any
* OS: Any
* CKEditor version: 4.6+
* Installed CKEditor plugins: pastefromword

HTML seen by browser (FF in that case) contains `margin-top`, `margin-bottom`. Unfortunately the PFW removes these styles and there is no way to prevent that. This is a huge problem since there is no way to prevent that.

**Editor should work in a way that non-configurable PFW allows a lot and configurable ACF can later be used to remove it. PFW should not remove styles which have significant impact on formatting.**

**HTML seen by browser:**
```
<p class="MsoListParagraphCxSpFirst" style="margin-top:0cm;margin-right:1.6pt;
margin-bottom:0cm;margin-left:36.0pt;margin-bottom:.0001pt;mso-add-space:auto;
text-align:justify;text-justify:inter-ideograph;text-indent:-36.0pt;line-height:
normal;mso-list:l1 level1 lfo1"><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;;mso-fareast-font-family:
Arial;mso-bidi-font-weight:bold" lang="EN-US"><span style="mso-list:Ignore">1.<span style="font:7.0pt &quot;Times New Roman&quot;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><span dir="LTR"></span><b><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;" lang="EN-US">Test1</span></b></p>

<p class="MsoListParagraphCxSpLast" style="margin-top:0cm;margin-right:1.6pt;
margin-bottom:0cm;margin-left:36.0pt;margin-bottom:.0001pt;mso-add-space:auto;
text-align:justify;text-justify:inter-ideograph;line-height:normal"><b><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;" lang="EN-US">&nbsp;</span></b></p>

<p class="MsoNormal" style="margin-left:36.0pt;text-align:justify;text-justify:
inter-ideograph;line-height:normal"><span style="font-size:12.0pt;
font-family:&quot;Arial&quot;,&quot;sans-serif&quot;;mso-fareast-font-family:&quot;Times New Roman&quot;" lang="EN-US">Test2</span></p>

<p class="MsoListParagraphCxSpFirst" style="text-align:justify;text-justify:inter-ideograph;
text-indent:-36.0pt;line-height:normal;mso-list:l1 level1 lfo1"><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;;mso-fareast-font-family:
Arial" lang="EN-US"><span style="mso-list:Ignore">2.<span style="font:7.0pt &quot;Times New Roman&quot;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><span dir="LTR"></span><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;" lang="EN-US">Test3</span><span style="font-size:12.0pt;font-family:&quot;Arial&quot;,&quot;sans-serif&quot;;mso-fareast-font-family:
Calibri" lang="EN-US"></span></p>
```

**HTML seen by the editor (Notice that margin-top and margin-bottom are removed from paragraph and list):**  
```
<ol>
	<li style="margin-right:2px; text-align:justify"><span style="font-size:11pt"><span style="text-justify:inter-ideograph"><span style="line-height:normal"><span style="font-family:&quot;Calibri&quot;,&quot;sans-serif&quot;"><b><span lang="EN-US" style="font-size:12.0pt"><span style="font-family:&quot;Arial&quot;,&quot;sans-serif&quot;">Test1</span></span></b></span></span></span></span></li>
</ol>

<p style="margin-right:2px; margin-left:48px; text-align:justify">&nbsp;</p>

<p style="margin-left:48px; text-align:justify; margin-bottom:13px"><span style="font-size:11pt"><span style="text-justify:inter-ideograph"><span style="line-height:normal"><span style="font-family:&quot;Calibri&quot;,&quot;sans-serif&quot;"><span lang="EN-US" style="font-size:12.0pt"><span style="font-family:&quot;Arial&quot;,&quot;sans-serif&quot;">Test2</span></span></span></span></span></span></p>

<ol start="2">
	<li style="text-align:justify"><span style="font-size:11pt"><span style="text-justify:inter-ideograph"><span style="line-height:normal"><span style="font-family:&quot;Calibri&quot;,&quot;sans-serif&quot;"><span lang="EN-US" style="font-size:12.0pt"><span style="font-family:&quot;Arial&quot;,&quot;sans-serif&quot;">Test3</span></span></span></span></span></span></li>
</ol>
```

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

Updated PFW filter to keep `margin-*:0` style when `config.pasteFromWord_keepEmptyMargins=true`.

Closes #2935 

**Important note: there is still ongoing discussion in #2935 until we fully agree if this solution is what we want this shouldn't be reviewed.**

### CODE REVIEW

**Comment by f1ames** [comment]
Rebased on latest `major`.

File: tests/plugins/pastefromword/manual/keepemptymargins.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+@bender-ckeditor-plugins: wysiwygarea, toolbar, undo, pastefromword, sourcearea
+
+1. Copy and paste content of following file into both editors:
+[Vertical_margin.docx](../generated/_fixtures/Vertical_margin/Vertical_margin.docx)
+1. Open source mode.
+
+## Expected
+
+- The first editor: the first `p` element has styles:
+```margin-right:0; margin-bottom:0; margin-left:0;```
```
Comment (f1ames) [inline tests/plugins/pastefromword/manual/keepemptymargins.md]: The styles in IE11 are a little different which might be misleading
Screen Shot 2019-04-11 at 18 05 07

File: plugins/pastefromword/filter/default.js [hunk 2]
```diff
@@ -2447,16 +2456,18 @@
...
+	function parseShorthandMargins( style, keepEmptyMargins ) {
 		var marginCase = style.margin ? 'margin' : style.MARGIN ? 'MARGIN' : false,
 			key, margin;
 		if ( marginCase ) {
 			margin = CKEDITOR.tools.style.parse.margin( style[ marginCase ] );
 			for ( key in margin ) {
 				var currMargin = margin[ key ];
-				// skip zeros.
-				if ( parseFloat( currMargin ) ) {
+				// We need to get rid of margins, unless they are allowed in config (#2935).
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply plugins/pastefromword/filter/default.js]: Lets stay with `zero` term as you've suggested name for config prop.

File: plugins/pastefromword/filter/default.js [hunk 3]
```diff
@@ -659,10 +661,16 @@
...
 					key = 'margin-' + key;
-					if ( !parseFloat( styles[ key ] ) ) {
-						delete styles[ key ];
+					if ( !( key in styles ) ) {
+						return;
+					}
+
+					var value = CKEDITOR.tools.convertToPx( styles[ key ] );
+					// We need to get rid of margins, unless they are allowed in config (#2935).
+					if ( value || keepEmptyMargins ) {
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply plugins/pastefromword/filter/default.js]: I'll rework `parseShorthandMargins` to preserve zeros, as it doesn't touch longhand margis, all the cleanup will stay in `normalizeMargins`.

**Comment by engineering-this** [comment]
I was sure I've added Safari fixtures, apparently I was wrong.

**Pull request merged**